internal/transport/handler_server.go

```diff
@@ -186,8 +186,9 @@ func (ht *serverHandlerTransport) WriteStatus(s *Stream, st *status.Status) erro
 	ht.writeStatusMu.Lock()
 	defer ht.writeStatusMu.Unlock()
 
+	s.updateHeaderSent()
```

Comment:
These calls need to happen outside of `ht.do` to make sure subsequent invocations fail immediately, even if at that point headers haven't not been fully written (but it's imminent).